_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+---
+layout: post
+title: The Road to the JIT
+tags: BEAM JIT HiPE
+author: Björn Gustavsson
+---
+
+As long as Erlang has existed, there has always been the need and the
+ambition to make it faster. This blog is a history lesson that
+outlines the major Erlang implementations and attempts to improve
+the performance of Erlang.
+
+### The Prolog interpreter
+
+The first version of Erlang was implemented in Prolog in 1986. That
+version of Erlang was too slow for creating real applications, but it
+was useful for finding out which features of the languages were useful
+and which were not. New languages features could be added or deleted
+in a matter of hours or days.
+
+It soon became clear that Erlang needed to be at least 40 times faster
+to be useful in real projects.
+
+### JAM (Joe's Abstract Machine)
+
+In 1989 JAM (Joe's Abstract Machine) was first
+implemented. [Mike Williams][mike] wrote the runtime system
+in C, [Joe Armstrong][joe] wrote the compiler, and
+[Robert Virding][robert] wrote the libraries.
+
+[mike]: http://www.erlang-factory.com/conference/ErlangUserConference2013/speakers/MikeWilliams
+[joe]: https://github.com/joearms
+[robert]: https://github.com/rvirding
+
+JAM was 70 times faster than the Prolog interpreter, but it turned out
+that still wasn't fast enough.
+
+### TEAM (Turbo Erlang Abstract Machine)
+
+Bogumil ("Bogdan") Hausman created TEAM (Turbo Erlang Abstract
+Machine). It compiled the Erlang code to C code, which was then
+compiled to native code using GCC.
+
+It was significantly faster than JAM for small projects.
+Unfortunately, compilation was very slow, and the code size of the
+compiled code was too big to make it useful for large projects.
+
+### BEAM (Bogdan's Erlang Abstract Machine)
+
+Bogumil Hausman next machine was called BEAM (Bogdan's Erlang Abstract
+Machine). It was a hybrid machine that could execute both native code
+(translated via C) and [threaded code] with an [interpreter]. That
+allowed customers to compile their time-critical modules to native code
+and all other modules to threaded BEAM code. The threaded BEAM in
+itself was faster than JAM code.
+
```

Comment:
I think it would be interesting if you can write something about why  threaded BEAM was faster than JAM.

Reply:
I think that is outside the scope of this blog post. If there is interest, I could write a separate blog post later.